### ISSUE TITLE
Wrong relocation emitted for dla in PIC mode

### ISSUE THREAD

**Issue by RichardsonAlex** [CTSRD-CHERI/llvm#188]
Clang emits a series of R_MIPS_HIGHEST R_MIPS_HIGHER, etc relocations for a 64 bit dla whereas GCC appears to use a %call16() relocation.

Interestingly BFD does not complain about this but lld does. I have a hacky workaround for this issue:  https://github.com/RichardsonAlex/cheribsd/commit/4372cd5dec8012fce4085d74ad25003c821e958a but I'm not sure that it's the right solution.

**Comment by RichardsonAlex** [comment]
I also checked the latest MipsAsmParser.cpp and it appears to be fixed for 32 bit, but not for 64 bit MIPS

**Comment by davidchisnall** [comment]
How does this work in general?  `dla` loads an address into a register, but `%call16` relocations are PC-relative offsets, so you can't use it to materialise an address.

**Comment by jrtc27** [comment]
1. `%call16` is GP-relative, so it would need to be `ld r, %call16(x)(gp)`
2. When assembling with `-KPIC`, GNU as seems to expand it to `ld r, %got_disp(x)(gp)`, because otherwise you would get the address of the PLT entry, which differs between DSOs.

**Comment by bsdjhb** [comment]
I'm having the same problem trying to build stock FreeBSD/mips with cheri-clang (shared libraries fail to link).  I've taken a blind stab at fixing this and seem to get the right looking things when I disassemble object files:

https://github.com/CTSRD-CHERI/llvm/compare/master...bsdjhb:mips64_pic_dla

**Comment by bsdjhb** [comment]
FYI, with the patches in the mips64_pic_dla branch (and a few other fixes), I've been able to build a FreeBSD/mips kernel + world with clang and old ld.bfd and run the resulting kernel + world under qemu.

**Comment by bsdjhb** [comment]
The dla patch was committed in http://llvm.org/viewvc/llvm-project?view=revision&revision=306831, so this can be closed once that is merged into CHERI llvm.

**Comment by bsdjhb** [comment]
Merged to CHERI llvm in c159742948721a0ba4abe3c320f8f250763b80bc
